src/types/redis_hash.cc

```diff
@@ -418,4 +472,208 @@ rocksdb::Status Hash::RandField(const Slice &user_key, int64_t command_count, st
+  std::vector<rocksdb::Status> statuses_vector;
+  statuses_vector.resize(sub_keys.size());
+  storage_->MultiGet(read_options, storage_->GetDB()->DefaultColumnFamily(), keys.size(), keys.data(),
+                     values_vector.data(), statuses_vector.data());
+
+  auto now = util::GetTimeStampMS();
+  for (size_t i = 0; i < keys.size(); i++) {
+    if (!statuses_vector[i].ok() && !statuses_vector[i].IsNotFound()) return statuses_vector[i];
+
+    // no such field exists
+    if (statuses_vector[i].IsNotFound()) {
+      ret->emplace_back(-2);
+      continue;
+    }
+
+    InternalKey sub_ikey(ns_key, fields[i], metadata.version, storage_->IsSlotIdEncoded());
+
+    // expire with a pass time
+    if (expire_ms <= now && !is_persist) {
+      batch->Delete(sub_ikey.Encode());
+      ret->emplace_back(2);
+      metadata.size -= 1;
+      continue;
+    }
+
+    auto value = values_vector[i].ToString();
+    uint64_t field_expire = 0;
+    decodeFieldAndTTL(metadata, &value, field_expire);
+    if (isMeetCondition(type, expire_ms, field_expire)) {
+      encodeFieldAndTTL(&value, expire_ms);
+      batch->Put(sub_ikey.Encode(), value);
+      if (is_persist && field_expire == 0) {
+        // for hpersist command, -1 if the field exists but has no associated expiration
+        ret->emplace_back(-1);
+      } else {
+        // 1 if expiration was updated or removed
+        ret->emplace_back(1);
+      }
+    } else {
+      ret->emplace_back(0);
+    }
+  }
+
+  // convert rest field encoding
+  if (!metadata.IsFieldExpirationEnabled()) {
+    metadata.field_encoding = HashSubkeyEncoding::WITH_TTL;
+
+    std::unordered_set<std::string_view> field_set;
+    for (auto field : fields) {
+      if (!field_set.emplace(field.ToStringView()).second) {
+        continue;
+      }
+    }
+
+    std::string prefix_key = InternalKey(ns_key, "", metadata.version, storage_->IsSlotIdEncoded()).Encode();
+    std::string next_version_prefix_key =
+        InternalKey(ns_key, "", metadata.version + 1, storage_->IsSlotIdEncoded()).Encode();
+
+    rocksdb::Slice upper_bound(next_version_prefix_key);
+    read_options.iterate_upper_bound = &upper_bound;
+
+    auto iter = util::UniqueIterator(storage_, read_options);
+    for (iter->Seek(prefix_key); iter->Valid() && iter->key().starts_with(prefix_key); iter->Next()) {
+      InternalKey sub_ikey(iter->key(), storage_->IsSlotIdEncoded());
+      auto value = iter->value().ToString();
+      if (field_set.find(sub_ikey.GetSubKey().ToStringView()) == field_set.end()) {
+        encodeFieldAndTTL(&value, 0);
+        batch->Put(sub_ikey.Encode(), value);
+      }
+    }
+  }
+
+  std::string bytes;
+  metadata.Encode(&bytes);
+  batch->Put(metadata_cf_handle_, ns_key, bytes);
+
+  return storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+
+rocksdb::Status Hash::TTLFields(const Slice &user_key, const std::vector<Slice> &fields, std::vector<int64_t> *ret) {
```

Comment:
please return an error status if the current hash object doesn't support field expiration.